rfcs/45-stageless.md

```diff
@@ -0,0 +1,1103 @@
+If neither system can write to the data the other reads, we *cannot tell* which order they ran in, and so any ordering constraint between them is pointless.
+This, of course, is equivalent to the two systems being compatible.
+
+Subtly, this **hypothetical compatibility** (also used to determined if system parameters conflict or if schedules are ambiguous) must be determined statically: on the basis of the filtered access to component and resource types (via `FilteredAccessSet<ComponentId>`).
+By contrast, **factual incompatibility** at the time of schedule execution is done based on the actual archetype-components that the systems access (via `Access<ArchetypeComponentId>`).
+Just like strict ordering constraints though, ordering constraints inferred on the basis of if-needed ordering constraints are respected at the time of schedule execution, even if there is no data conflict at the time the systems are being run.
+
+If-needed ordering constraints are intended to behave exactly like strict ordering constraints (unless interior mutability or other strangeness is involved), and so they are transitive.
+In order to achieve this, the following algorithm is used:
+
+1. For each system that is at the start of an if-needed constraint:
+   1. Walk down the tree of if-needed ordering constraints and identify all downstream systems.
+   2. Add an if-needed ordering constraint between the root system and each downstream system.
+2. For each if-needed ordering constraint:
+   1. If the two systems are compatible, remove the constraint.
+   2. If they are incompatible, promote it to a strict ordering constraint.
+3. Deduplicate all strict ordering constraints.
+   1. Each pair of systems can only have one ordering constraint between them.
+      1. At this stage, all ordering constraints have been converted to strict ordering constraints or removed.
+   2. Any edges that are implied by the transitive property can be removed.
+
+### Atomic groups
+
+For all systems that are part of the same atomic system group:
+
+- When a system in an atomic group completes, its write-locks and read-locks are downgraded to **virtual locks**.
+- Virtual locks are treated as if they were real locks for all systems outside of the atomic group.
+- Virtual locks are ignored by systems in the same atomic group.
+- If the atomic group is **coherent**, virtual locks are removed once all systems in the atomic group have finished with it.
+- If the atomic group is **isolated**, virtual locks are held until all systems in the atomic group complete.
+
+This can be modelled as:
+
+```rust
+enum DataLock{
+   Read,
+   Write,
+   VirtualRead(Box<dyn SystemLabel>),
+   VirtualWrite(Box<dyn SystemLabel>),
+}
+```
+
+Multiple virtual locks on the same data can exist at once, and membership in an atomic group is *not* transitive.
+Suppose we have two atomic groups, `GroupX = {A, B}` and `GroupY = {B, C}`.

```

Comment:
My instincts say we shouldn't allow overlapping atomic groups that don't have a subset superset relationship. Like an atomic group is supposed to make multiple systems look like one system to outside systems and overlapping systems just don't compute for me.

That said this is mostly an instinctual reaction and I haven't thought through the consequences of having overlapping atomic groups.

Reply:
I'll move this comment over to the new RFC.